control-plane/addMissingIds.cpp

```diff
@@ -0,0 +1,214 @@
+        newAnnos->add(new IR::Annotation("id", annoExprs));
+        valueSet->annotations = newAnnos;
+    }
+    return valueSet;
+}
+
+const IR::P4Action* MissingIdAssigner::preorder(IR::P4Action* action) {
+    if (ControlPlaneAPI::isHidden(action)) {
+        return action;
+    }
+    CHECK_NULL(symbols);
+    auto anno = ControlPlaneAPI::getIdAnnotation(action);
+    if (!anno) {
+        const auto* annos = action->getAnnotations();
+        auto* newAnnos = annos->clone();
+        IR::Vector<IR::Expression> annoExprs;
+        auto symbolId = symbols->getId(
+            ControlPlaneAPI::P4RuntimeSymbolType::ACTION(), action);
+        const auto* idConst =
+            new IR::Constant(new IR::Type_Bits(ID_BIT_WIDTH, false), symbolId);
+        annoExprs.push_back(idConst);
+        newAnnos->add(new IR::Annotation("id", annoExprs));
+        action->annotations = newAnnos;
+    }
+
+    auto* paramList = new IR::ParameterList();
+    for (size_t symbolId = 0; symbolId < action->parameters->size();
+         ++symbolId) {
+        auto* param = action->parameters->getParameter(symbolId)->clone();
+        auto anno = ControlPlaneAPI::getIdAnnotation(param);
+        if (!anno) {
+            const auto* annos = param->getAnnotations();
+            auto* newAnnos = annos->clone();
+            IR::Vector<IR::Expression> annoExprs;
+            const auto* idConst = new IR::Constant(
+                new IR::Type_Bits(ID_BIT_WIDTH, false), symbolId + 1);
+            annoExprs.push_back(idConst);
+            newAnnos->add(new IR::Annotation("id", annoExprs));
+            param->annotations = newAnnos;
+        }
+        paramList->parameters.push_back(param);
+    }
+    action->parameters = paramList;
+    return action;
+}
+MissingIdAssigner::MissingIdAssigner(
+    ReferenceMap* refMap, TypeMap* typeMap,
+    const ControlPlaneAPI::P4RuntimeArchHandlerBuilderIface& archBuilder)
+    : refMap(refMap), typeMap(typeMap), archBuilder(archBuilder) {
+    CHECK_NULL(typeMap);
+    CHECK_NULL(refMap);
+    setName("MissingIdAssigner");
+}
+
+MissingIdAssigner::MissingIdAssigner(
+    ReferenceMap* refMap, TypeMap* typeMap,
+    const ControlPlaneAPI::P4RuntimeSymbolTable* symbols,
+    const ControlPlaneAPI::P4RuntimeArchHandlerBuilderIface& archBuilder)
+    : refMap(refMap),
+      typeMap(typeMap),
+      symbols(symbols),
+      archBuilder(archBuilder) {
+    CHECK_NULL(typeMap);
+    CHECK_NULL(refMap);
+    CHECK_NULL(symbols);
+    setName("MissingIdAssigner");
+}
+
+AddMissingIdAnnotations::AddMissingIdAnnotations(
+    ReferenceMap* refMap, TypeMap* typeMap,
+    const ControlPlaneAPI::P4RuntimeArchHandlerBuilderIface* archBuilder) {
+    CHECK_NULL(refMap);
+    CHECK_NULL(typeMap);
+
+    passes.push_back(new ControlPlaneAPI::ParseP4RuntimeAnnotations());
+    passes.push_back(new MissingIdAssigner(refMap, typeMap, *archBuilder));
+    // Need to refresh types after we have updated some expressions with
```

Comment:
the rule we use is the opposite: if a pass needs types, it should run first typeChecking.
So this should not be necessary - whatever comes after will do it if needed.